portal/trigger_states/empro_states.py

```diff
@@ -0,0 +1,392 @@
+     https://promowiki.movember.com/display/ISS/Trigger+Logic
+
+    :param qnr: QuestionnaireResponse to process as "current"
+    :return: None
+
+    """
+    try:
+        # first, confirm state transition is allowed - raises if not
+        ts = users_trigger_state(qnr.subject_id)
+        sm = EMPRO_state(ts)
+
+        if ts.state == 'processed' and override_state:
+            # go around state machine, setting directly when requested
+            current_app.logger.debug(
+                f"override trigger_state transition from {ts.state} "
+                f"to 'inprocess'")
+            ts.state = 'inprocess'
+
+        # typical flow, processing was triggered before SDC handoff
+        # if launched from testing or some catch-up task, initiate now
+        if ts.state != "inprocess":
+            current_app.logger.debug(
+                "evaluate_triggers(): trigger_state transition from "
+                f"{ts.state} to 'inprocess'")
+            enter_user_trigger_critical_section(user_id=qnr.subject_id)
+            # confirm local vars picked up state change
+            assert ts.state == 'inprocess'
+
+        # bring together and evaluate available data for triggers
+        dm = DomainManifold(qnr)
+        ts.triggers = dm.eval_triggers()
+        ts.questionnaire_response_id = qnr.id
+
+        # transition and persist state
+        sm.processed_triggers()
+        current_app.logger.debug(
+            "record state change to 'processed' from "
+            f"evaluate_triggers() for {qnr.subject_id}")
+        ts.insert(from_copy=True)
+
+        return ts
+
+    except (TransitionNotAllowed, LockTimeout) as e:
+        current_app.logger.exception(e)
+        raise e
+
+    finally:
+        # All done, release semaphore for this user
+        critical_section = TimeoutLock(key=EMPRO_LOCK_KEY.format(
+            user_id=qnr.subject_id))
+        critical_section.__exit__(None, None, None)
+
+
+def fire_trigger_events():
+    """Typically called as a celery task, fire any pending events
+
+    After questionnaire responses and resulting observations are evaluated,
+    a celery job call lands here to seek out and execute any appropriate
+    events from the user's trigger state.  Said rows will be in the
+    'processed' state.
+
+    Actions are recorded in trigger_states.triggers and the row's state
+    is transitioned to 'triggered' or should no further action be necessary,
+    'resolved'.
+
+    """
+    from ..models.user import User
+    from .empro_messages import patient_email, staff_emails
+    from ..date_tools import FHIR_datetime
+
+    def send_n_report(em, context, record):
+        """Send email, append success/fail w/ context to record"""
+        result = {'context': context, 'timestamp': FHIR_datetime.now()}
+        try:
+            em.send_message()
+            result['email_message_id'] = em.id
+            record.append(result)
+        except SMTPRecipientsRefused as e:
```

Comment:
Should we handle that exception in the [logging module](https://github.com/uwcirg/truenth-portal/blob/develop/portal/logs.py#L61-L64)?

Reply:
As we're catching here - shouldn't ever hit.  No, that clause in the logging module is a backdoor to break out of a cmd line execution.